smolyay/surrogate.py

```diff
@@ -117,430 +72,694 @@ def data(self):
 
-    @domain.setter
-    def domain(self, value):
-        value = numpy.array(value, ndmin=2)
-        if value.ndim != 2 or value.shape[1] != 2:
-            raise TypeError('Domain must be dim x 2')
-        self._domain = value
-        self._reset_grid()
+        Parameters
+        ----------
+        function: callable
+            Function to be approximated.
+        """
+        data = [function(x) for x in X]
+        self.fit(X, data)
 
-    @property
-    def grid(self):
-        """:class:`IndexGrid` Sampling grid."""
-        if self._grid is None:
-            self._grid = self.grid_generator(self.dimension)
-        return self._grid
+    @abc.abstractmethod
+    def fit(self, X, y=None):
+        """Fit surrogate's components (basis functions) to data.
 
-    @property
-    def grid_generator(self):
-        """:class:`IndexGridGenerator`: Sampling grid generator."""
-        return self._grid_generator
+        Parameters
+        ----------
+        X : list
+            points that are sampled
 
-    @grid_generator.setter
-    def grid_generator(self, value):
-        if not isinstance(value, IndexGridGenerator):
-            raise TypeError('Grids must be IndexGridGenerator')
-        self._grid_generator = value
-        self._reset_grid()
+        y : list
+            function at grid points.
 
-    @property
-    def points(self):
-        """numpy.ndarray: Points to be sampled."""
-        if self._points is None:
-            self._points = self._mapdomain(
-                self.grid.points, [[-1, 1]]*self.dimension, self._domain)
-        return self._points
+        Returns
+        -------
+        self : object
+            BaseSurrogate class instance
 
-    def gradient(self, x):
-        """Take the derivative of the surrogate at a given point.
+        Raises
+        ------
+        ValueError
+            Input must lie in domain of surrogate.
+        """
 
-        The derivative of the surrogate can be obtained at a given
-        point inside the approximation domain after the surrogate
-        has been constructed and all of its coefficients have been computed.
+    @abc.abstractmethod
+    def predict(self, X):
+        """Evaluate surrogate at a given input.
 
         Parameters
         ----------
-        x: list (for dimension > 1), or float (for dimension = 1)
-            Input.
+        X: array-like with shape (n_samples, num_dimensions) or list of object
+            Points at which the model is evaluated
 
         Returns
         -------
-        list (for dimension > 1) or float (for dimension = 1)
-            Derivative of the surrogate in all dimensions.
+        ndarray of shape (n_samples,) or (n_samples, n_targets)
+            Surrogate output at x.
 
         Raises
         ------
-        ValueError
+        RuntimeError
             For surrogate to be evaluated, function needs to be trained.
+        ValueError
```

Comment:
I'm not seeing `fit_gradient` or `predict_gradient` in this base class. Is that intentional?

Reply:
Yes. One of the reasons why the base surrogate class exists instead of just handling everything it handles with the SetProductSurrogate is if there could potentially be a surrogate that's just a container for a sklearn surrogate model. 

(The reason that could potentially justify this container class is the fact that the container class for the GaussianProcessRegressor scales the domain of the inputs to something like [-1, 1], and I've found that doing this improves the GaussianProcessRegressor with it has to model a function with a large domain.) 

GaussianProcessRegressors don't have gradients or predict gradient, so I can't guarantee all children of the base surrogate will be capable of implementing those.